File: 프로그래머스/lv3/43163. 단어 변환/단어 변환.js
function solution(begin, target, words) {
  const checkOneDiff = (word1, word2) => {
    let countDiff = 0;
    for (let i = 0; i < word1.length; i++) {
      if (word1[i] !== word2[i]) countDiff++;
      if (countDiff > 1) return false;
    }
    return countDiff === 0 ? false : true;
  };
  const newWords = words.unshift(begin);
  let targetIdx = words.findIndex((el) => el === target);
  if (targetIdx === -1) return 0;

  const n = words.length;
  const visited = Array.from({ length: n }).fill(0);
  const bfs = (start) => {
    let q = [];
    q.push(start);
    while (q.length) {
      const current = q.shift();
      for (let i = 1; i < n; i++) {
        const next = i;
        if (visited[next] === 0 && checkOneDiff(words[current], words[next])) {
          q.push(next);
          visited[next] = visited[current] + 1;
        }
      }
    }
  };
  bfs(0);
  return visited[targetIdx];
}

File: 프로그래머스/lv3/43163. 단어 변환/단어 변환.py
import sys
sys.setrecursionlimit(10**6)

def solution(begin, target, words):
    l = len(begin)
    def check(word1, word2):
        count = 0
        for i in range(l):
            if word1[i] != word2[i]:
                count += 1
                if count >= 2: return False
        if count == 1: return True
        return False
            
    start = -1
    answer = 0
    visited = [0 for _ in range(len(words)+1)]
    def dfs(node, depth):
        nonlocal answer
        if words[node] == target: answer = depth
        for i, word in enumerate(words):
            cur_word = ''
            if node == -1: cur_word = begin
            else: cur_word = words[node]
            if not visited[i+1] and check(word, cur_word):
                visited[i+1] = 1
                dfs(i, depth+1)
                visited[i+1] = 0
    visited[0] = 1
    dfs(-1, 0)
    return answer